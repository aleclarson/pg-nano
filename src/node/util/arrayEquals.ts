export function arrayEquals(
  a: unknown[] | null | undefined,
  b: unknown[] | null | undefined,
): boolean {
  if (!a) {
    return !b
  }
  if (!b) {
    return false
  }
  return (
    a === b ||
    (a.length === b.length && a.every((val, index) => Object.is(val, b[index])))
  )
}
