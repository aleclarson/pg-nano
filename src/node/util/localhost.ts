/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2014 Joseph Werle
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import fs from 'node:fs'
import net from 'node:net'

export function isLocalHost(host: string): boolean {
  if (host === '::1' || host === 'localhost') {
    return true
  }

  if (process.env.DEV_CONTAINER === 'pg-nano' && host === 'postgres') {
    return true
  }

  if (net.isIP(host)) {
    for (const line of readHostsFile()) {
      if (/^(127\.0\.0\.1|::1)/.test(line.trim())) {
        const hosts = line.split(/\s+/)
        if (hosts.includes(host)) {
          return true
        }
      }
    }
    return false
  }

  // 127.0.0.1 - 127.255.255.254
  const parts = host.split('.').map(Number)
  if (parts.length !== 4) {
    return false
  }
  if (parts[0] !== 127) {
    return false
  }
  if (parts[1] > 255 || parts[1] < 0) {
    return false
  }
  if (parts[2] > 255 || parts[2] < 0) {
    return false
  }
  if (parts[3] > 254 || parts[3] < 1) {
    return false
  }
  return true
}

function readHostsFile(): string[] {
  let configs: string
  try {
    configs = fs.readFileSync('/etc/hosts', 'utf-8')
  } catch {
    return []
  }
  return configs.split('\n')
}
