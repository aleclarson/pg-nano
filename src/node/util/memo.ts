export interface MemoOptions<Fn extends (...args: any[]) => any> {
  cache?: Record<string, ReturnType<Fn>>
  key?: (...args: Parameters<Fn>) => string
}

/**
 * Creates a memoized function. The returned function will only
 * execute the source function when no value has previously been
 * computed.
 *
 * @see https://radashi-org.github.io/reference/curry/memo
 * @example
 * ```ts
 * const calls: number[] = []
 * const fib = memo((x: number) => {
 *   calls.push(x)
 *   return x < 2 ? x : fib(x - 1) + fib(x - 2)
 * })
 *
 * fib(10) // 55
 * fib(10) // 55
 * // calls === [10]
 *
 * fib(11) // 89
 * // calls === [10, 11]
 * ```
 */
export function memo<Fn extends (...args: any[]) => any>(
  func: Fn,
  options: MemoOptions<Fn> = {},
): Memoized<Fn> {
  const { key: keyFunc, cache = {} } = options

  return function callWithMemo(...args) {
    const key = keyFunc ? keyFunc(...args) : JSON.stringify({ args })
    if (key in cache) {
      return cache[key]
    }
    const result = func(...args)
    cache[key] = result
    return result
  } as Memoized<Fn>
}

export type Memoized<Fn extends (...args: any[]) => any> =
  ReturnType<Fn> extends infer Result
    ? (...args: Parameters<Fn>) => Result
    : never
