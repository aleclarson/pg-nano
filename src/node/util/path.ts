import path from 'node:path'

export function cwdRelative(p: string) {
  const rel = path.relative(process.cwd(), p)
  return rel.startsWith('..') ? rel : './' + rel
}
