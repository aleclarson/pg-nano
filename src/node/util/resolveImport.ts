export function resolveImport(specifier: string) {
  // Once Vite supports import.meta.resolve, we can remove the require.resolve
  // fallback. See: https://github.com/vitejs/vite/discussions/15871
  let resolvedPath: string
  if (typeof import.meta.resolve === 'function') {
    resolvedPath = new URL(import.meta.resolve(specifier)).pathname
  } else {
    resolvedPath = require.resolve(specifier)
  }

  return resolvedPath
}
