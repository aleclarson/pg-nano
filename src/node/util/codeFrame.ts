import { cwdRelative } from './path.js'

export function appendCodeFrame(
  error: Error,
  errorPosition: number,
  query: string,
  stmtLine = 1,
  stmtFile?: string,
) {
  const errorLine = query.slice(0, errorPosition).split('\n').length - 1
  const errorColumn =
    errorPosition - (1 + query.lastIndexOf('\n', errorPosition - 1))

  const queryLines = (query + ';').split('\n')
  const startLine = Math.max(0, errorLine - 2)
  const endLine = Math.min(queryLines.length - 1, errorLine + 3)
  const width = String(stmtLine + endLine).length

  let output = ''

  for (let line = startLine; line <= endLine; line++) {
    output += `${line === errorLine ? '> ' : '  '}${String(stmtLine + line).padStart(width)} | ${queryLines[line]}\n`

    if (line === errorLine) {
      output += ' '.repeat(3 + width) + `| ${' '.repeat(errorColumn - 1)}^\n`
    }
  }

  error.message =
    error.message.replace(/\nLINE \d+: [\s\S]+$/, '') +
    '\n\n' +
    output +
    (stmtFile
      ? '\n    at ' +
        cwdRelative(stmtFile) +
        ':' +
        (stmtLine + errorLine) +
        ':' +
        errorColumn
      : '')
}
