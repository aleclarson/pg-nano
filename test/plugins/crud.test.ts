import { createProject, dedent, resetPublicSchema } from '../util.js'

const sql = dedent

describe('@pg-nano/plugin-crud', () => {
  beforeEach(resetPublicSchema)

  test('basic cases', async () => {
    const project = await createProject(
      {
        'sql/schema.sql': sql`
          CREATE TABLE "user" (
            id bigint PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
            name text NOT NULL,
            tags text[] DEFAULT '{}'
          );

          CREATE TABLE "book" (
            sku text PRIMARY KEY,
            title text NOT NULL
          );
        `,
      },
      {
        plugins: ['@pg-nano/plugin-crud'],
      },
    )

    await project.generate()
    expect(project.generatedFiles).toMatchSnapshot()

    type Schema =
      typeof import('../__fixtures__/@pg-nano+plugin-crud__basic_cases/sql/schema.js')

    const client = await project.importClient<Schema>()

    // Create a record
    expect(await client.createUser({ name: 'John' })).toEqual({
      id: 1n, // <== generated id
      name: 'John',
      tags: [], // <== default value
    })

    // Get a record
    expect(await client.getUser(1)).toEqual({
      id: 1n,
      name: 'John',
      tags: [],
    })

    // Update a record
    expect(await client.updateUser(1, { name: 'Jonny' })).toEqual({
      id: 1n,
      name: 'Jonny',
      tags: [],
    })

    // Verify update
    expect(await client.getUser(1)).toEqual({
      id: 1n,
      name: 'Jonny',
      tags: [],
    })

    // Delete a record
    expect(await client.deleteUser(1)).toEqual(true)

    // Verify deletion
    expect(await client.getUser(1)).toEqual(null)

    // Upsert a non-existent record
    expect(
      await client.upsertBook({
        sku: 'abc',
        title: 'The Art of Computer Programming',
      }),
    ).toEqual({
      sku: 'abc',
      title: 'The Art of Computer Programming',
    })

    // Verify upsert
    expect(await client.getBook('abc')).toEqual({
      sku: 'abc',
      title: 'The Art of Computer Programming',
    })

    // Upsert an existing record
    expect(
      await client.upsertBook({
        sku: 'abc',
        title: 'The Art of Computer Programming Vol. 1',
      }),
    ).toEqual({
      sku: 'abc',
      title: 'The Art of Computer Programming Vol. 1',
    })

    // Verify upsert
    expect(await client.getBook('abc')).toEqual({
      sku: 'abc',
      title: 'The Art of Computer Programming Vol. 1',
    })
  })

  test('get - null', async () => {
    const project = await createProject(
      {
        'sql/schema.sql': sql`
          CREATE TABLE foo (
            id integer PRIMARY KEY
          );
        `,
      },
      {
        plugins: ['@pg-nano/plugin-crud'],
      },
    )

    await project.generate()

    type Schema =
      typeof import('../__fixtures__/@pg-nano+plugin-crud__get_null/sql/schema.js')

    const client = await project.importClient<Schema>()

    expect(await client.getFoo(1)).toEqual(null)
  })

  test('create - pk conflict', async () => {
    const project = await createProject(
      {
        'sql/schema.sql': sql`
          CREATE TABLE "grocery_list" (
            list_id integer,
            item_id integer,
            name text NOT NULL,
            PRIMARY KEY (list_id, item_id)
          );
        `,
      },
      {
        plugins: ['@pg-nano/plugin-crud'],
      },
    )

    await project.generate()

    type Schema =
      typeof import('../__fixtures__/@pg-nano+plugin-crud__create_pk_conflict/sql/schema.js')

    const client = await project.importClient<Schema>()

    // Create first item successfully
    expect(
      await client.createGroceryList({ listId: 1, itemId: 1, name: 'Salad' }),
    ).toEqual({
      listId: 1,
      itemId: 1,
      name: 'Salad',
    })

    // Attempt to create second item with same primary key
    await expect(() =>
      client.createGroceryList({ listId: 1, itemId: 1, name: 'Cookies' }),
    ).rejects.toThrowErrorMatchingInlineSnapshot(`
      [PgResultError: ERROR:  duplicate key value violates unique constraint "grocery_list_pkey"
      DETAIL:  Key (list_id, item_id)=(1, 1) already exists.
      CONTEXT:  SQL statement "INSERT INTO "public"."grocery_list"
          VALUES ($1[1]::int4, $1[2]::int4, $1[3])
          RETURNING *"
      PL/pgSQL function create_grocery_list(text[]) line 6 at SQL statement
      ]
    `)
  })
})
