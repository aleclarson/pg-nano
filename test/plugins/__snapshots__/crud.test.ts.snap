// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@pg-nano/plugin-crud > basic cases 1`] = `
{
  "sql/schema.ts": "/* BEWARE: This file was generated by pg-nano. Any changes you make will be overwritten. */
import { bindQueryRowOrNull, bindQueryValue } from 'pg-nano'
import * as t from './typeData.js'

export type User = {
  id: number
  name: string
  tags?: string[]
}
export declare namespace User {
  type InsertParams = {
    id?: number
    name: string
    tags?: string[]
  }
  type UpsertParams = InsertParams & {
    id: number
  }
}

export declare namespace createUser {
  type Params = [User.InsertParams]
  type Result = User
}

export const createUser = /* @__PURE__ */ bindQueryRowOrNull<createUser.Params, createUser.Result>("create_user", $ => $.arity(1, 1).mapInput(0, t.insert_mapper(t.user)).returnsRecord())

export declare namespace deleteUser {
  type Params = { id: number }
  type Result = boolean
}

export const deleteUser = /* @__PURE__ */ bindQueryValue<deleteUser.Params, deleteUser.Result>("delete_user", $ => $.arity(1, 1).namedArgs(["id"]))

export declare namespace getUser {
  type Params = { id: number }
  type Result = User
}

export const getUser = /* @__PURE__ */ bindQueryRowOrNull<getUser.Params, getUser.Result>("get_user", $ => $.arity(1, 1).namedArgs(["id"]).returnsRecord())

export declare namespace updateUser {
  type Params = [number, Partial<User>]
  type Result = User
}

export const updateUser = /* @__PURE__ */ bindQueryRowOrNull<updateUser.Params, updateUser.Result>("update_user", $ => $.arity(2, 2).mapInput(1, t.update_mapper).returnsRecord())

export declare namespace upsertUser {
  type Params = [User.UpsertParams]
  type Result = User
}

export const upsertUser = /* @__PURE__ */ bindQueryRowOrNull<upsertUser.Params, upsertUser.Result>("upsert_user", $ => $.arity(1, 1).mapInput(0, t.insert_mapper(t.user)).returnsRecord())
",
  "sql/typeData.ts": "/* BEWARE: This file was generated by pg-nano. Any changes you make will be overwritten. */
import * as t from './typeData.js'

export { defineArrayMapper as array, defineRowMapper as row } from 'pg-nano'

// Field mappers
export { insert_mapper } from "@pg-nano/plugin-crud/field-mappers"
export { update_mapper } from "@pg-nano/plugin-crud/field-mappers"

// Composite types
export const user = /* @__PURE__ */ t.row(["id","name","tags"])
",
}
`;
